### PR TITLE
Filtering

### PR DESCRIPTION
**Pull request by vmarsi**

### CODE REVIEW

**Comment by zsvizi** [comment]
Create a separate class for `filter_by_gauge,` `filter_multiple_gauges`, and `filter_by_water_level.` Since this class will contain static methods, no constructor will be needed.

File: src/analysis.py [hunk 1]
```diff
@@ -298,3 +298,136 @@ def func(j_graph, start_nodes, end_nodes):
                                                                   end_station=section[1], func=func))
         flood_map.add_weighted_edges_from(ebunch_to_add=flood_map_edges)
         return flood_map
+
+    @staticmethod
+    def filter_by_gauge(graph,
```
Comment (zsvizi) [inline src/analysis.py]: Annotate arguments and return value and add docstring to this method.

File: src/analysis.py [hunk 2]
```diff
@@ -298,3 +298,136 @@ def func(j_graph, start_nodes, end_nodes):
...
+                    edges.remove(edge)
+
+        g = nx.DiGraph()
+        g.add_nodes_from(nodes)
+        g.add_edges_from(edges)
+
+        return g
+
+    @staticmethod
+    def filter_multiple_gauges(graph,
```
Comment (zsvizi) [inline src/analysis.py]: Annotate arguments and return value and add docstring to this method.

File: src/analysis.py [hunk 3]
```diff
@@ -298,3 +298,136 @@ def func(j_graph, start_nodes, end_nodes):
...
+                    edges.remove(edge)
+
+        g = nx.DiGraph()
+        g.add_nodes_from(nodes)
+        g.add_edges_from(edges)
+
+        return g
+
+    @staticmethod
+    def filter_by_water_level(graph,
```
Comment (zsvizi) [inline src/analysis.py]: Annotate arguments and return value and add docstring to this method.

File: src/analysis.py [hunk 4]
```diff
@@ -298,3 +298,136 @@ def func(j_graph, start_nodes, end_nodes):
...
+        comps = list(nx.weakly_connected_components(graph))
+        edges = graph.edges()
+        comps_copy = comps.copy()
+        edges = list(edges)
+        edges_copy = edges.copy()
+
+        for comp in comps_copy:
+            if not any(gauge == elem for elem in [i[0] for i in [list(ele) for ele in list(comp)]]):
+                comps.remove(comp)
+        for i in range(len(comps)):
```
Comment (zsvizi) [inline src/analysis.py]: There is a duplicated code fragments warning here. Create a function, which follows this logic and can be applied to the other methods.

**Comment by zsvizi** [comment]
Remove `flood_wave_trial.ipynb` from the repository, leave the `notebooks` folder "empty" (it will contain `.gitignore`), upload to notebook to this Google Drive folder (https://drive.google.com/drive/folders/1kTq3nWuezXOnLtrov3I3U9ZS657kWOeN?usp=sharing) and add a link to the notebook in the `README.md`

File: src/plotter.py [hunk 5]
```diff
@@ -274,6 +290,46 @@ def create_nan_graph(self, min_date: str, max_date: str):
             nan_dates = gauge_data[str(gauge)].index[gauge_data[str(gauge)].apply(np.isnan)].strftime("%Y-%m-%d")
 
             for date in nan_dates:
-                nan_graph.add_node(node_for_adding=(gauge, date))
+                nan_graph.add_node(node_for_adding=(gauge, date), color="green")
 
         return nan_graph
+
+    def get_subgraph_by_level_group(self, directed_graph: nx.DiGraph):
```
Comment (zsvizi) [inline src/plotter.py]: Annotate return value, add docstring.

File: src/plotter.py [hunk 6]
```diff
@@ -274,6 +290,46 @@ def create_nan_graph(self, min_date: str, max_date: str):
...
+        all_connected_subgraphs = []
```
Comment (zsvizi) [inline src/plotter.py]: This method might have a better place in `analysis.py`.

File: src/plotter.py [hunk 7]
```diff
@@ -274,6 +290,46 @@ def create_nan_graph(self, min_date: str, max_date: str):
...
+                    print(SG.nodes)
+                    all_connected_subgraphs.append(SG)
+        print(all_connected_subgraphs)
+
+    def color_and_label(self, directed_graph: nx.DiGraph, folder_name: str):
+        colors = [""] * len(self.positions)
+        labels = {}
+        g = open(os.path.join(PROJECT_PATH, "data", "level_groups_fontos.json"))
+        level_groups = json.load(g)
+        for gauge in self.gauges:
```
Comment (zsvizi) [inline src/plotter.py:302]: This is technically a part for data handling/engineering. It looks more reasonable to move this part to the flood wave detection part, and store the outcome of this snippet in a generated data file (JSON/CSV). Then this method will only load that file and do its job.

File: src/plotter.py (hunk 5 above)
Comment (vmarsi) [reply src/plotter.py]: Szerintem ezt a metódust első próbálkozásaim között írtam, amikor még nem tudtam, hogy van olyan beépített függvény, hogy `nx.is_weakly_connected()`. Ahogy elnézem, valami hasonlót próbáltam írni kézzel. Ez sehol máshol nincs használva a repo-ban (szerencsére, mert nem is működik), úgyhogy ezt törlöm inkább.

File: src/analysis.py (hunk 4 above)
Comment (zsvizi) [reply src/analysis.py]: This problem occurs still in `filtering.py` (LOC81 vs. LOC147).

File: data/level_groups_fontos.json [hunk 8]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (zsvizi) [inline data/level_groups_fontos.json]: Remove this file, since `Dataloader` downloads it.

File: src/flood_wave_detector.py [hunk 9]
```diff
@@ -82,6 +83,8 @@ def find_vertices(self) -> None:
         The end result is saved to 'PROJECT_PATH/generated/find_vertices' folder.
         :return:
         """
+        g = open(os.path.join(PROJECT_PATH, "data", "level_groups_fontos.json"))
```
Comment (zsvizi) [inline src/flood_wave_detector.py]: I would put this file load to `find_local_maxima`, so `find_local_maxima` will not have a `level_group` argument.

File: src/flood_wave_handler.py [hunk 10]
```diff
@@ -295,40 +295,49 @@ def remove_components_not_including_start_or_end_station(
...
+            peak_list_new.append(tuple((peak_list[i][0], peak_list[i][1], color)))
+        return peak_list_new
 
     @staticmethod
     def clean_dataframe_for_getting_peak_list(
             local_peak_values: np.array,
             gauge_data: pd.DataFrame,
-            reg_number: str
+            reg_number: str,
+            level_group: float
```
Comment (zsvizi) [inline src/flood_wave_handler.py]: You can remove this argument, since you do not use it in the method.

File: src/plotter.py [hunk 11]
```diff
@@ -1,7 +1,9 @@
 from datetime import datetime, timedelta
 from typing import Union
 import os
+import json
 
+import itertools
```
Comment (zsvizi) [inline src/plotter.py]: Unused import statement.

File: src/flood_wave_handler.py [hunk 12]
```diff
@@ -328,7 +331,7 @@ def clean_dataframe_for_getting_peak_list(
         peaks.info()
         peaks.index = pd.to_datetime(peaks.index).strftime('%Y-%m-%d')
         peaks.info()
-        peaks = peaks.drop(columns="Date") #      .set_index(peaks.index.strftime('%Y-%m-%d'))
+        peaks = peaks.drop(columns="Date")  # .set_index(peaks.index.strftime('%Y-%m-%d'))
         peaks[reg_number] = peaks[reg_number].astype(float)
         return peaks
```
Comment (zsvizi) [inline src/flood_wave_handler.py:337]: Since this class is static, do not instantiate in the notebook.

**Review by zsvizi — CHANGES_REQUESTED** [review]

**Review by zsvizi — APPROVED** [review]

**Pull request merged**